File: java-tomcat/src/main/java/com/ederfmatos/javatomcat/log/Logger.java
package com.ederfmatos.javatomcat.log;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Consumer;
import lombok.Getter;

public sealed interface Logger permits DefaultLogger {
    static Logger create(Class<?> clazz) {
        return new DefaultLogger(clazz);
    }

    @Getter
    class Context {
        private String message;
        private final Map<String, Object> params = new HashMap<>();

        public Context withMessage(String message) {
            this.message = message;
            return this;
        }

        public void withParam(String key, Object value) {
            this.params.put(key, value);
        }
    }

    void info(Consumer<Context> log);
}

File: kotlin-tomcat/src/main/kotlin/com/ederfmatos/kotlintomcat/log/Logger.kt
package com.ederfmatos.kotlintomcat.log

import kotlin.reflect.KClass

sealed interface Logger {
    fun info(messageBlock: Context.() -> Unit)

    data class Context(var message: String, val params: MutableMap<String, Any>) {
        fun withMessage(message: String): Context {
            this.message = message
            return this
        }

        fun withParam(key: String, value: Any): Context {
            this.params[key] = value
            return this
        }
    }

    companion object {
        fun create(clazz: KClass<*>): Logger {
            return KotlinLogger(clazz)
        }
    }
}